src/vset.c

```diff
@@ -0,0 +1,1249 @@
+        break;
+    }
+    case VSET_BUCKET_SINGLE: {
+        entry = vsetBucketSingle(*set);
+        expiry = getExpiry(entry);
+        if (expiry > now)
+            return NULL;
+        break;
+    }
+    case VSET_BUCKET_VECTOR: {
+        entry = pvGet(vsetBucketVector(*set), 0);
+        expiry = getExpiry(entry);
+        if (expiry > now)
+            return NULL;
+        break;
+    }
+    case VSET_BUCKET_HT: {
+        hashtableIterator iter;
+        hashtableInitIterator(&iter, vsetBucketHashtable(*set), 0);
+        assert(hashtableNext(&iter, &entry));
+        hashtableResetIterator(&iter);
+        expiry = getExpiry(entry);
+        if (expiry > now)
+            return NULL;
+        break;
+    }
+    default:
+        panic("Unknown volatile set bucket type in vsetNext");
+    }
+    if (delete)
+        assert(vsetRemoveEntry(set, getExpiry, entry, expiry));
+    return entry;
+}
+
+void *vsetPopExpired(vset *set, vsetGetExpiryFunc getExpiry, mstime_t now) {
+    return vsetGetFirstExpired(set, getExpiry, now, true);
+}
+
+void *vsetFirstExpired(vset *set, vsetGetExpiryFunc getExpiry, mstime_t now) {
+    return vsetGetFirstExpired(set, getExpiry, now, false);
+}
+
+int vsetNext(vsetIterator *it, void **entryptr) {
+    vsetBucket *bucket = it->bucket;
+    int bucket_type = vsetBucketType(bucket);
+    int ret = 0;
+    switch (bucket_type) {
+    case VSET_BUCKET_NONE:
+        return vsetBucketNext_NONE(it, entryptr);
+        break;
+    case VSET_BUCKET_RAX:
+        return vsetBucketNext_RAX(it, entryptr);
+        break;
+    case VSET_BUCKET_SINGLE:
+        ret = vsetBucketNext_SINGLE(it, entryptr);
+        break;
+    case VSET_BUCKET_VECTOR:
+        ret = vsetBucketNext_VECTOR(it, entryptr);
+        break;
+    case VSET_BUCKET_HT:
+        ret = vsetBucketNext_HASHTABLE(it, entryptr);
+        break;
+    default:
+        panic("Unknown volatile set bucket type in vsetNext");
+    }
+    if (ret == 0) {
+        /* continue iterating the parent bucket */
+        it->iteration_state = vsetBucketType(it->parent_bucket);
+        it->bucket = it->parent_bucket;
+        return vsetNext(it, entryptr);
+    }
+    return ret;
+}
+
+void vsetStart(vset *set, vsetIterator *it) {
+    it->iteration_state = VSET_BUCKET_NONE; /*lets start by going to the first bucket. */
+    it->bucket = *set;
+    it->bucket_ts = -1;
+    it->parent_bucket = vsetBucketSetNone(it->parent_bucket);
+}
+
+void vsetStop(vsetIterator *it) {
+    int bucket_type = vsetBucketType(it->bucket);
+    int parent_bucket_type = vsetBucketType(it->parent_bucket);
+    if (parent_bucket_type == VSET_BUCKET_RAX)
+        raxStop(&it->riter);
+    if (bucket_type == VSET_BUCKET_HT)
+        hashtableResetIterator(&it->hiter);
+}
+
+vset *createVolatileSet(vset *set) {
+    *set = vsetBucketSetNone(*set);
+    return set;
+}
+
+void freeVolatileSet(vset *set) {
+    if (!(*set)) return;
+    freeVsetBucket(*set);
+}
```

Comment:
These too should use the vset prefix?

Since they are not allocating/freeing the `set` itself, I guess they should be called init and cleanup/clear etc. 

The init function should probably return void like other init functions.

The clear/cleanup/release function could set the set to empty to prevent double free, etc. Is that a good idea?

```suggestion
void vsetInit(vset *set) {
    *set = vsetBucketFromNone();
}

void vsetClear(vset *set) {
    if (!(*set)) return;
    freeVsetBucket(*set);
    *set = vsetBucketFromNone();
}
```

Reply:
I like your proposal. will change it